### ISSUE TITLE
replace ts-node with esbuild?

### ISSUE THREAD

**Issue by abcd-ca** [knex/knex#5198]
I just replaced ts-node with esbuild for locally running my Node.js (express.js, TypeScript project) server and it's way faster now. When I remove ts-node from package.json, Knex CLI doesn't run anymore - I gather it has a handful of "registers" that it looks for and is looking for babel as a default, having not found any. If I re-add ts-node, the Knex CLI it runs again. Is there a way to get Knex to run with [esbuild-register](https://github.com/egoist/esbuild-register)?

**Comment by mauricedoepke** [comment]
The code for parsing typescript is here:
https://github.com/knex/knex/blob/8a50dd7462d40cfa122a233f7fcada24ea39d3ae/bin/cli.js#L106

Knex uses rechoir and intperpret for doing so.
I guess the easiest way to implement this would be to allo users supplying custom interpret configs to rechoir.

**Comment by mauricedoepke** [comment]
I guess @abcd-ca use case would already be solved if knex would update "interpret" to the newest version. The newest version supports esbuild out of the box and should automatically select it when ts-node is not available.

**Comment by nour-dev** [comment]
> I guess @abcd-ca use case would already be solved if knex would update "interpret" to the newest version. The newest version supports esbuild out of the box and should automatically select it when ts-node is not available.


this did not work (I mean updating interpret to latest version)

**Comment by lucapasquale** [comment]
Any updates here? I've been having some issues with `ts-node` that doesn't occur with `esbuild`, so I would like to do that change

**Comment by kibertoad** [comment]
PR would be welcome!

**Comment by mauricedoepke** [comment]
@kibertoad From my understanding just the interpret version needs to be bumped. There is already a pr from dependabot:
https://github.com/knex/knex/pull/5244

I personally patched my local knex version with the swc/register code from the newest interpret and it works for me. So I highly suspect this will enable esbuild as well.

But I would recommend releasing this as a new major version as the newest interpret version removed some outdated ways of loading typescript. See:

https://github.com/gulpjs/interpret/releases/tag/v3.0.0

**Comment by mauricedoepke** [comment]
@lucapasquale You should be able to force knex to use the newest interpret version manually in the mean time via:
yarn: https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions/
pnpm: https://pnpm.io/package_json#pnpmoverrides
npm: https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides

Just use this to force it to interpret 3.1.1

Then you have to install esbuild-register: https://github.com/egoist/esbuild-register as esbuild alone will not be enough, I think.

Then it should work.

**Comment by noor-tg** [comment]
@mauricedoepke I just use ts-node with swc

**Comment by mercmobily** [comment]
Marking this as a big, that will be fixed once we update all of the dependencies (coming soon!)

**Comment by mercmobily** [comment]
BTW... Great minds think alike! You're absolutely right that ts-node is slow and there are better alternatives. We're actually going one better than esbuild - in #5800 we're migrating to tsx, which uses esbuild under the hood but adds a lot of Node.js-specifi intelligence on top.

  tsx is perfect for Knex because:
  - It's as fast as esbuild (because it uses esbuild internally)
  - It handles all the Node.js module resolution edge cases
  - It works with both ESM and CommonJS seamlessly
  - It doesn't require any configuration

  Once #5800 is complete, you'll get the speed benefits you want without needing to keep ts-node around just for Knex. In the  meantime, you can use:

```
  NODE_OPTIONS='--import tsx/esm' knex migrate:latest
```

This gives you the performance boost today!

**Comment by mercmobily** [comment]
Actually I am closing this in favour of #5800 -- let me know if you think that's not a good idea. #5800 is high priority and will get done as soon as we are finished sorting out the issue list!
